### DISCUSSION TITLE
REST API Deprecated

### DISCUSSION THREAD

**Discussion by iqtedar123** [timdorr/tesla-api#762] [General]
Looks like they are deprecating the REST API.
https://github.com/teslamotors/vehicle-command?s=09
Seems we have to register our apps with Tesla now.
Thoughts?

**Comment by Urkman** [comment]
I tried to create a developer account and register my app, but I only get an error: "unable to onboard app"...

**Reply by vagh141** [reply]
I wonder gore strict they will be to approve apps. Will they approve most apps as long as they don't violate their terms?

**Reply by SylvainGa** [reply]
I got up to that page. Not sure what to enter since my app doesn't use any servers on my side.

![Screenshot_20231012-182234](https://github.com/timdorr/tesla-api/assets/49377866/38622c06-d495-4263-ba7a-f3c2f015b698)

**Reply by iqtedar123** [reply]
I think most of us will have to shutdown our projects since we were relying on the REST API directly.
Looks like Tesla wants developers to register and pay for the API (most likely companies like Hertz etc will enroll) since they want to launch an app store.
Sad news indeed.

**Reply by SylvainGa** [reply]
Shit. Tesla's third party apps where one thing that made Tesla stands out. Sucks that they want to kill it 🙁

**Reply by tomhollander** [reply]
@SylvainGa 

> I got up to that page. Not sure what to enter since my app doesn't use any servers on my side.

For the Origin URL, this needs to be an actual website which you own, as you'll need to upload keys there as part of the account registration process. I used a free Azure storage account for mine. Your actual app does not need to integrate with this URL. 

The Reply URL can be a fake URL like blah://blah for a native client app. Needs to be real for a web app. 

I don't know what the Return URL does.

**Reply by SeBsZ** [reply]
Anyone else getting the "unable to onboard app" error after entering everything? I'm pretty sure my URLs are valid.
![image](https://github.com/timdorr/tesla-api/assets/5457531/189370f9-780b-460a-b0f3-7cc099b57899)

What can I do to continue?

**Reply by westerlund** [reply]
@SeBsZ same

**Comment by purcell-lab** [comment]
> I think most of us will have to shutdown our projects since we were relying on the REST API directly.
> Looks like Tesla wants developers to register and pay for the API (most likely companies like Hertz etc will enroll) since they want to launch an app store.
> Sad news indeed.

No need to start spreading FUD until we have the details.  A documented and supported API from Tesla is a good thing and much better than trying to constantly reverse engineer an unofficially supported API, which can be switched off at a moments notice.

OAuth authentication is an industry standard and there are lots of authentication providers out there, including the ability to host your own OpenSource OAuth server.  The legacy API currently uses OAuth authentication so the apps already have a level of OAuth compatability.

Yes existing apps will need to migrate to the new framework, but the fact that Tesla have provided a documented RestAPI and proxy server which appears to have more functionality than the current undocumented API does give me hope. Looks like the developer tier is still free, so let's not scare the horses too early.

Some folks have this working already in their dev environments:
https://github.com/timdorr/tesla-api/issues/763#issuecomment-1758933555

**Reply by iqtedar123** [reply]
This is a discussion post. So a little bit of FUD should be allowed. 😅
Just sharing what other publications are reporting:
https://www.notateslaapp.com/news/1653/tesla-creates-official-apis-for-third-party-services-to-start-charging-for-usage?s=09

Totally agree that having official support for the API is great.
However it does come at the cost of indie developers who now have to apply for the developer program and will most likely be rejected (for various reasons).

**Comment by BJReplay** [comment]
> I got up to that page. Not sure what to enter since my app doesn't use any servers on my side.
> 
Those aren't servers, per se, but you'll need to do a bit of work for OAuth.

Their expectation seems to be that OAuth requests will come from a domain with a redirect and return.

It might be as simple as setting up DNS entries to get past this screen, and using those same DNS entries in your app.  

You might need these to exist when you register.

I am a few days away from testing this (just had shoulder surgery, so typing with my wrong hand), but I'm the same situation as you, but I think I know enough  about OAuth to be dangerous.

Let's keep this discussion going.

**Reply by Rockster160** [reply]
I'm a back end web developer- I'm quite familiar with the ins and outs of Oauth. Don't have much experience with the 3rd party Oauth services, but I'm happy to answer questions or help out getting an actual system set up.
It is very likely you need a DNS and simple server running in order to get it to work.

**Reply by SylvainGa** [reply]
Thanks

 Does the server need a static IP address or can a dynamic DNS service like duckdns work?

**Reply by Rockster160** [reply]
Yep! Dynamic IP addresses should work with Oauth in general- although I'm not sure about Tesla's API directly. I'll say it's very unlikely they have an issue with dynamic ips, but you never know. 😂

**Comment by tomhollander** [comment]
For anyone using my [TeslaAuth](https://github.com/tomhollander/TeslaAuth) library for C#/.NET, I've managed to get this library to work with the Fleet API. The changes haven't been merged yet and testing/feedback would be appreciated. Once it's confirmed to work for people besides myself, I'll merge and push to Nuget.

https://github.com/tomhollander/TeslaAuth/pull/36

**Reply by BJReplay** [reply]
@tomhollander Tom would you consider updating the version of Newtonsoft.Json that you're using?

The 13 series release came out Mar 23, 2021, and I'm pretty sure I upgraded as it fixed an issue I had parsing JSON from some external source that was fixed in that release.

When I upgraded beyond 12.0.3, I ran into an issue as yours requires this version `<PackageReference Include="Newtonsoft.Json" Version="12.0.3" />`, so I had an issue using the NuGet version of your package with the older version of Newtonsoft.Json.

I forked your package, and updated my fork to use 13.0.1 (I haven't updated since, though 13.0.3 is now out for more than six months), and compiled it in.

**Reply by tomhollander** [reply]
I might just bite the bullet and migrate to System.Text.Json.

**Reply by tomhollander** [reply]
3.0.1 published

**Reply by jonasman** [reply]
btw how did you configure your return_uri in developer account? im getting an error all the time when trying to get the auth token from tesla using my uri as tesla requests like this: https://victorious-cliff-xxxxxxx.azurestaticapps.net/auth/callback

**Reply by tomhollander** [reply]
The redirect URI can be whatever you want but the value you pass at runtime needs to match one of the options you specified when registering your account. This is the redirect URI; the return URI is different and I haven't figured out what that's for.

**Reply by jonasman** [reply]
nothing works here.
im trying to call this api: https://auth.tesla.com/oauth2/v3/authorize
with the exact redirect_uri as i placed in the developer account: `https://victorious-cliff-xxxxxxx.azurestaticapps.net` no luck
adding `https://victorious-cliff-xxxxxxx.azurestaticapps.net/auth/callback` no luck
both URL encoded strings

the error:
We don't recognize this redirect URI/URL. Please use the redirect URI/URL found in your app details.

**Reply by jonasman** [reply]
the partner register API is also giving an error. damn

**Reply by tomhollander** [reply]
@jonasman Are you using my library? Works fine for me. 
<img width="519" alt="image" src="https://github.com/timdorr/tesla-api/assets/10659986/f999c766-65f9-4ee8-b779-c5a8b9bf035b">

**Reply by jonasman** [reply]
@tomhollander no, im trying to update my own swift library.
The prob is `https://auth.tesla.com/oauth2/v3/authorize` doesnt like my redirect uri that i configured in developer.tesla.com, no idea why

**Reply by jonasman** [reply]
ok i added `testapp://test` as a redirect uri to a new app and it worked maybe tesla wont allow all the urls to be the same..
also it wont allow `https://google.com`

**Reply by Urkman** [reply]
I also made some progress here. The "authorize" request works fine and the callback url is called with a "code" inside.
But when I then try to create the token using the "token" request I always get a bad_request (400). But without any error message.
Any ideas on this? 

<img width="1165" alt="Bildschirmfoto 2023-10-30 um 13 43 21" src="https://github.com/timdorr/tesla-api/assets/94297/ddac2551-88f2-4469-8feb-f679e7b1fd3f">

**Reply by jonasman** [reply]
looks correct to me. maybe some problem in the url encoding? 
Unless your hidden values are incorrect?

**Reply by Urkman** [reply]
It works when I remove the header "--header 'Content-Type: application/x-www-form-urlencoded' \"...
Strange...

**Reply by Urkman** [reply]
So, one step further :) 
I registered my app using the "partner_accounts" endpoint with a "partner token". This worked fine.

But if I want to call a vehicle endpoint, it gives me "Account **** must be registered please see https://developer.tesla.com/docs/fleet-api#register".

Did you manage to call a vehicle endpoint of the new API?

<img width="1433" alt="Bildschirmfoto 2023-10-30 um 17 55 44" src="https://github.com/timdorr/tesla-api/assets/94297/cdb7bcc8-b17b-4a8f-8943-cfbaf8ef86c7">

**Reply by Urkman** [reply]
Ok, fount it... It was just the url. I needed "https://fleet-api.prd.eu.vn.cloud.tesla.com", so "EU" and "NA" :)
So, now it works for me :)

**Reply by leovolwahsen** [reply]
> I also made some progress here. The "authorize" request works fine and the callback url is called with a "code" inside. But when I then try to create the token using the "token" request I always get a bad_request (400). But without any error message. Any ideas on this?
> 
> <img alt="Bildschirmfoto 2023-10-30 um 13 43 21" width="1165" src="https://private-user-images.githubusercontent.com/94297/279059809-ddac2551-88f2-4469-8feb-f679e7b1fd3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3NDUwODYsIm5iZiI6MTcyMTc0NDc4NiwicGF0aCI6Ii85NDI5Ny8yNzkwNTk4MDktZGRhYzI1NTEtODhmMi00NDY5LThmZWItZjY3OWU3YjFmZDNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIzVDE0MjYyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1OTU5Y2RjNWU5YmQ3N2RhMTg3MDM1ZGFkOTY0YTRjN2YwYTFhMmRmZjE1Nzg1ZmM3MDhjMTMxYWFmMjBkYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hcFVD6Iuh4-iDp6FbCA3ioMthQWg2Q7AhfyeK0MR2eo">

@Urkman could you help me understand how you got the value for code? The other parameter I manage to get...

**Comment by tomhollander** [comment]
> For anyone using my [TeslaAuth](https://github.com/tomhollander/TeslaAuth) library for C#/.NET, I've managed to get this library to work with the Fleet API. The changes haven't been merged yet and testing/feedback would be appreciated. Once it's confirmed to work for people besides myself, I'll merge and push to Nuget.

This has now been released as v3.0.0 of the library (it still also works for the Owner API)

**Comment by BJReplay** [comment]
> I've managed to get this library to work with the Fleet API.

That's great news, I look forward to testing this.  I'll be a few days away (post shoulder surgery - I don't really want to re-factor just typing with my wrong hand).

I've been using a source code version for a while as I ran into an issue with the  Newtonsoft.Json JSON library that I used in my app that was incompatible with the version in your library (possibly due to me needing (?wanting) a newer version, or a .Net Framework version issue, so will be keen to report back on this, too.  I note from reviewing the PR that you considered moving away from it, but didn't.

**Comment by jonasman** [comment]
The old api might not be dead as Tesla wrote this in the command api repo:

> Pre-2021 Model S and X vehicles will continue to support the full REST API, and do not support the new protocol.

**Reply by samosss** [reply]
I have a Model S 2019 build (but upgraded from HW2.5 to HW3 computer), and it _seems_ like REST API is _not_ supported on mine (Australia, if that makes any difference)

**Comment by FlorianR-Hub** [comment]
Hello ! 

Does anyone know how to integrate the [Vehicle Command SDK](https://github.com/teslamotors/vehicle-command) into an Android (or even Flutter) application?

The SDK has been developed using the Go language. It's normally possible to compile Go code so that it can be used by mobile applications, but it looks a bit tricky.

Tesla also offers a proxy, to convert REST API calls to their new vehicle-command protocol.
But if I've understood correctly, this requires you to have your own proxy, which implies additional infrastructure costs. Not ideal for small developers...

Are there any other solutions?

**Reply by LandonMoss** [reply]
Add the SDK Dependency:

If the Vehicle Command SDK supports Flutter, you can add it as a dependency in your pubspec.yaml file.
yaml
Copy code
dependencies:
  vehicle_command_sdk: ^version
Integrate Native Code (if necessary):

If the SDK is written in native code (Java/Kotlin for Android), you might need to create a Flutter plugin that acts as a bridge between Flutter and the native SDK.
Initialize and Use the SDK:

In your Flutter code, initialize the SDK and use its functions. This often involves calling native methods through the Flutter plugin.
dart
Copy code
import 'package:vehicle_command_sdk/vehicle_command_sdk.dart';

void main() {
  VehicleCommandSDK.init(apiKey);
}
Handle SDK Events:

If the SDK triggers events or callbacks, make sure to handle them appropriately in your Flutter code.

**Reply by FlorianR-Hub** [reply]
Thank you for your answer.

However, these are the obvious solutions, and there is currently only an SDK developed in GO.
I'm not aware of any public Flutter/Dart dependencies developed by Tesla or anyone else.
For native code, same thing, you'd have to do it yourself.

The point of my previous message was to find out if there were other solutions already possible at the time that I'd missed.

In the meantime, there's been some news from the Tesla SDK on Github. They've added documentation explaining how the protocol works to send commands to Tesla vehicles.

I don't see Tesla coming up with a dependency for Android or Flutter applications, so I think there are only 2 possibilities for those who want to integrate this SDK in mobile apps :
- wait for someone to make a dependency available
- write the dependency yourself

**Reply by SylvainGa** [reply]
> Hello ! 
> 
> Does anyone know how to integrate the [Vehicle Command SDK](https://github.com/teslamotors/vehicle-command) into an Android (or even Flutter) application?
> 
> The SDK has been developed using the Go language. It's normally possible to compile Go code so that it can be used by mobile applications, but it looks a bit tricky.
> 
> Tesla also offers a proxy, to convert REST API calls to their new vehicle-command protocol.
> But if I've understood correctly, this requires you to have your own proxy, which implies additional infrastructure costs. Not ideal for small developers...
> 
> Are there any other solutions?
> 

The method I chose, which made some of the users of my (free to use) app unhappy is to strip the Tesla code and replace it with the equivalent Tessie API commands. Of course, you need a (paid) Tessie account to use their API (which made some users unhappy), but it's simple to use and supported by the new Tesla API.

I'm not planning on installing/supporting a publicly available proxy server that could make me responsible if something/someone gets hacked through it because I failed to update it/or through a zero day exploit. No thank you, not for me.
